src/sentry/models/dynamicsampling.py

```diff
@@ -0,0 +1,258 @@
+    condition = models.TextField()
+    sample_rate = models.FloatField(default=0.0)
+    start_date = models.DateTimeField(default=timezone.now)
+    end_date = models.DateTimeField()
+    num_samples = models.IntegerField()
+    condition_hash = models.CharField(max_length=40)
+
+    @property
+    def external_rule_id(self) -> int:
+        """
+        Returns the external rule id
+
+        For external users, i.e. Relay, we need to shift the ids since the slot we
+        have allocated starts at the offset specified in RESERVED_IDS.
+        """
+        return self.rule_id + CUSTOM_RULE_START
+
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_customdynamicsamplingrule"
+        indexes = [
+            # get active rules for an organization
+            models.Index(fields=["organization"], name="org_idx", condition=Q(is_active=True)),
+            # get expired rules (that are still marked as active)
+            models.Index(fields=["end_date"], name="end_date_idx", condition=Q(is_active=True)),
+            # find active rules for a condition
+            models.Index(
+                fields=["condition_hash"], name="condition_hash_idx", condition=Q(is_active=True)
+            ),
+        ]
+
+    @staticmethod
+    def get_rule_for_org(
+        condition: Any, organization_id: int
+    ) -> Optional["CustomDynamicSamplingRule"]:
+        """
+        Returns an active rule for the given condition and organization if it exists otherwise None
+
+        Note: There should not be more than one active rule for a given condition and organization
+        This function doesn't verify this condition, it just returns the first one.
+        """
+        condition_hash = get_condition_hash(condition)
+        rules = CustomDynamicSamplingRule.objects.filter(
+            organization_id=organization_id,
+            condition_hash=condition_hash,
+            is_active=True,
+            end_date__gt=timezone.now(),
+        )
+        for rule in rules:
+            return rule
+        return None
+
+    @staticmethod
+    def update_or_create(
+        condition: Any,
+        start: datetime,
+        end: datetime,
+        project_ids: Sequence[int],
+        organization_id: int,
+        num_samples: int,
+        sample_rate: float,
+    ) -> "CustomDynamicSamplingRule":
+        with transaction.atomic(router.db_for_write(CustomDynamicSamplingRule)):
+
+            # check if rule already exists for this organization
+            existing_rule = CustomDynamicSamplingRule.get_rule_for_org(condition, organization_id)
+
+            if existing_rule is not None:
+                # we already have an active rule for this condition and this organization
+                # update the expiration date and ensure that our projects are included
+                existing_rule.end_date = max(end, existing_rule.end_date)
+                existing_rule.num_samples = max(num_samples, existing_rule.num_samples)
```

Comment:
Is `num_samples` a static number on how many samples we want to collect? If so, it's a bit weird to not have last writer wins semantics on rule update. Maybe I am missing some business logic here.

Reply:
The problem with max is that we are losing one number, so a sum of the samples would be more representative, ofc it might result in oversampling.